backend/s3: crash in s3.Get

terraform v0.9.6

terraform apply  - 

 main.tf in attached ZIP.

has worked before, but I got confused about an S3 backend. Did a destroy on everything, then started again. Crashes every time, even after reboot.

OS X 10.12.5. iTerm2. All AWS in eu-west-2 (London)


[Archive.zip](https://github.com/hashicorp/terraform/files/1038447/Archive.zip)


### Affected Resource(s)
Please list the resources as a list, for example:
- aws_instance
- dnsimple_record

If this issue appears to affect multiple resources, it may be an issue with Terraform's core, so please mention this.

### Terraform Configuration Files
terraform.tfstate attached

### Debug Output
Please provider a link to a GitHub Gist containing the complete debug output: https://www.terraform.io/docs/internals/debugging.html. Please do NOT paste the debug output in the issue; just paste a link to the Gist.

### Panic Output
`crash.log`. in zip file

### Expected Behavior
What should have happened?
should create a simple ASG & ELB, from a book.

### Actual Behavior
Crash every time



### Steps to Reproduce

1. `terraform apply`

### Important Factoids
All very standard, following book 'terraform up and running'

Panic output: 
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x10 pc=0x46f62aa]

goroutine 158 [running]:
github.com/hashicorp/terraform/backend/remote-state/s3.(*RemoteClient).Get(0xc420053340, 0x11, 0x11, 0x18)
        /private/tmp/terraform-20170525-36317-scr3s4/terraform-0.9.6/src/github.com/hashicorp/terraform/backend/remote-state/s3/client.go:64 +0x11a
github.com/hashicorp/terraform/state/remote.(*State).RefreshState(0xc420781d20, 0x100ed72, 0xc420554140)
        /private/tmp/terraform-20170525-36317-scr3s4/terraform-0.9.6/src/github.com/hashicorp/terraform/state/remote/state.go:33 +0x37
github.com/hashicorp/terraform/state.(*BackupState).RefreshState(0xc42057b890, 0x5e21a2f, 0x7)
        /private/tmp/terraform-20170525-36317-scr3s4/terraform-0.9.6/src/github.com/hashicorp/terraform/state/backup.go:21 +0x33
github.com/hashicorp/terraform/backend/local.(*Local).context(0xc4205c5680, 0xc42009e420, 0x2, 0x2, 0x4, 0xc420554140, 0x2)
        /private/tmp/terraform-20170525-36317-scr3s4/terraform-0.9.6/src/github.com/hashicorp/terraform/backend/local/backend_local.go:31 +0x97
github.com/hashicorp/terraform/backend/local.(*Local).opApply(0xc4205c5680, 0x88caa20, 0xc4205540c0, 0xc42009e420, 0xc420473ce0)
        /private/tmp/terraform-20170525-36317-scr3s4/terraform-0.9.6/src/github.com/hashicorp/terraform/backend/local/backend_apply.go:50 +0x1fd
github.com/hashicorp/terraform/backend/local.(*Local).(github.com/hashicorp/terraform/backend/local.opApply)-fm(0x88caa20, 0xc4205540c0, 0xc42009e420, 0xc420473ce0)
        /private/tmp/terraform-20170525-36317-scr3s4/terraform-0.9.6/src/github.com/hashicorp/terraform/backend/local/backend.go:245 +0x52
github.com/hashicorp/terraform/backend/local.(*Local).Operation.func1(0xc4205c5680, 0xc4209e6eb0, 0xc4209e6ea0, 0x88caa20, 0xc4205540c0, 0xc42009e420, 0xc420473ce0)
        /private/tmp/terraform-20170525-36317-scr3s4/terraform-0.9.6/src/github.com/hashicorp/terraform/backend/local/backend.go:265 +0x9a
created by github.com/hashicorp/terraform/backend/local.(*Local).Operation
        /private/tmp/terraform-20170525-36317-scr3s4/terraform-0.9.6/src/github.com/hashicorp/terraform/backend/local/backend.go:266 +0x192
```

--- a/backend/remote-state/s3/client.go
+++ b/backend/remote-state/s3/client.go
@@ -58,11 +58,19 @@ func (c *RemoteClient) Get() (payload *remote.Payload, err error) {
 			return nil, err
 		}
 
+		// If the remote state was manually removed the payload will be nil,
+		// but if there's still a digest entry for that state we will still try
+		// to compare the MD5 below.
+		var digest []byte
+		if payload != nil {
+			digest = payload.MD5
+		}
+
 		// verify that this state is what we expect
 		if expected, err := c.getMD5(); err != nil {
 			log.Printf("[WARNING] failed to fetch state md5: %s", err)
-		} else if len(expected) > 0 && !bytes.Equal(expected, payload.MD5) {
-			log.Printf("[WARNING] state md5 mismatch: expected '%x', got '%x'", expected, payload.MD5)
+		} else if len(expected) > 0 && !bytes.Equal(expected, digest) {
+			log.Printf("[WARNING] state md5 mismatch: expected '%x', got '%x'", expected, digest)
 
 			if testChecksumHook != nil {
 				testChecksumHook()
@@ -74,7 +82,7 @@ func (c *RemoteClient) Get() (payload *remote.Payload, err error) {
 				continue
 			}
 
-			return nil, fmt.Errorf(errBadChecksumFmt, payload.MD5)
+			return nil, fmt.Errorf(errBadChecksumFmt, digest)
 		}
 
 		break

--- a/backend/remote-state/s3/client_test.go
+++ b/backend/remote-state/s3/client_test.go
@@ -167,6 +167,8 @@ func TestRemoteClient_clientMD5(t *testing.T) {
 		"lock_table": bucketName,
 	}).(*Backend)
 
+	createS3Bucket(t, b.s3Client, bucketName)
+	defer deleteS3Bucket(t, b.s3Client, bucketName)
 	createDynamoDBTable(t, b.dynClient, bucketName)
 	defer deleteDynamoDBTable(t, b.dynClient, bucketName)
 
@@ -263,8 +265,8 @@ func TestRemoteClient_stateChecksum(t *testing.T) {
 		t.Fatal(err)
 	}
 
-	// put the old state in place of the new, without updating the checksum
-	if err := client2.Put(oldState.Bytes()); err != nil {
+	// put an empty state in place to check for panics during get
+	if err := client2.Put([]byte{}); err != nil {
 		t.Fatal(err)
 	}
 
@@ -278,7 +280,18 @@ func TestRemoteClient_stateChecksum(t *testing.T) {
 	consistencyRetryTimeout = 0
 	consistencyRetryPollInterval = 0
 
-	// fetching the state through client1 should now error out due to a
+	// fetching an empty state through client1 should now error out due to a
+	// mismatched checksum.
+	if _, err := client1.Get(); !strings.HasPrefix(err.Error(), errBadChecksumFmt[:80]) {
+		t.Fatalf("expected state checksum error: got %s", err)
+	}
+
+	// put the old state in place of the new, without updating the checksum
+	if err := client2.Put(oldState.Bytes()); err != nil {
+		t.Fatal(err)
+	}
+
+	// fetching the wrong state through client1 should now error out due to a
 	// mismatched checksum.
 	if _, err := client1.Get(); !strings.HasPrefix(err.Error(), errBadChecksumFmt[:80]) {
 		t.Fatalf("expected state checksum error: got %s", err)

